File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.go
//https://leetcode.com/problems/unique-morse-code-words/discuss/138396/0ms-golang-solution
func uniqueMorseRepresentations(words []string) int {
    table := []string{".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."}
    
    result := make(map[string]bool)
    for _, w := range words {
        var res string
        for _, r  := range w {
            res += table[unicode.ToLower(r)-'a']
        }
        result[res] = true
    }
    
    return len(result)
}

//More space efficient
import "strings"

func uniqueMorseRepresentations(words []string) int {
    uniq := make(map[string]int)
    
    for _, word := range words {
        uniq[Morsify(word)]++
    }
    
    return len(uniq)
}

var morseTable = [...]string{
    ".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--..",
}

func Morsify(word string) string {
    sb := strings.Builder{}
    for _, c := range word {    
        sb.WriteString(morseTable[c - 97])
    }
    return sb.String()
}

File: Solutions/804_Unique_Morse_Code_Words/804_Unique_Morse_Code_Words.rs
//https://leetcode.com/problems/unique-morse-code-words/discuss/214221/Rust-Solution
use std::collections::HashSet;
impl Solution {
    pub fn unique_morse_representations(words: Vec<String>) -> i32 {
        
        let morse_codes = vec![".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."];
        
        let mut morse_set = HashSet::new();
        for word in &words{
            let mut temp_str : String = "".into();
            let mut chars = word.chars();
            let mut each_char = chars.next();
            loop {
                match each_char {
                    Some(x) => temp_str.push_str(morse_codes[(x as u8 - 97) as usize]),
                    None => break,
                }
             each_char = chars.next();
            }
            morse_set.insert(temp_str);
        }
        return morse_set.len() as i32;
    }
    
}

//More space efficient
use std::collections::HashSet;

impl Solution {
    // Don't get too invested with lifetime.
    pub const morse: [&'static str; 26] = [
        ".-","-...","-.-.","-..",".","..-.","--.","....","..",
        ".---","-.-",".-..","--","-.","---",".--.","--.-",".-.",
        "...","-","..-","...-",".--","-..-","-.--","--.."
    ];
        
    pub fn unique_morse_representations(words: Vec<String>) -> i32 {
        let mut h: HashSet<String> = HashSet::with_capacity(words.len());
        
        for word in words {
            let code = word
                        .chars()
                        .map(|x| Solution::morse[(x.to_digit(36).expect("") - 10) as usize])
                        .collect::<Vec<&str>>()
                        .join("");
            h.insert(code);
        }
        h.len() as i32
    }
}